### DISCUSSION TITLE
Server callback on multiple emit()

### DISCUSSION THREAD

**Discussion by seatedscribe** [miguelgrinberg/python-socketio#1129] [Q&A] [answered]
So I have this async server and multiple clients connected. From time to time I need to distribute data over different clients so I multiple-emit to different rooms, one per client. Each client process this data and return its result, which is used by the server to compose the final result. 

How is the server callback managed in this scenario wrt to "thread" safety? Is there any async-related issue I must consider, e.g. when two or more client replies trigger the execution of this callback on the server at the same time? I guess callback is not guaranteed to be executed atomically out of the box, but I may be wrong. 

Thanks
Enrico

**Comment by miguelgrinberg** [comment, marked as answer]
Callbacks are supported only when addressing a single client. You can emit to your clients one by one in a loop and can bypass this limitation, and then it would be up to your callbacks to figure out which client is responding. You can use `functools.partial()` to bind each callback to the client it belongs to, for example.

**Reply by seatedscribe** [reply]
Yes I am emitting to multiple clients in a loop already. No "broadcast"here, just a 1-to-1 (per client) emit pattern, if your concern was related to performing a single emit but involving multiple clients.

Alas, I cannot have separate callbacks as I don't know in advance how many clients are available.

My fear is that the (single) callback may be triggered from a second client while the first call has not completed yet. The callback basically is modifying a local dict, filling an unique key (sid) with received data, but I am aware that modifying dict concurrently is not advisable even if data does not overlap.

Should I put a lock in the callback body, for serenity of mind, or should I ignore this altogether as it's already handled at lower level (socketio-asyncio)?

Hope I made it clearer.
Thanks in advance.

**Reply by miguelgrinberg** [reply]
> I cannot have separate callbacks as I don't know in advance how many clients are available.

To be clear, I did not mean different functions. You can use the same callback function for all your clients as long as you come up with a method to identify the client in each callback invocation. I suggested `functools.partial()` as a solution to that, but if you have another solution, then you should be fine too.

> My fear is that the (single) callback may be triggered from a second client while the first call has not completed yet.

This package does not attempt to serialize callbacks, so it is possible to have multiple callbacks running concurrently. Concurrent access to a dict is sometimes safe and sometimes unsafe, depending on how the application does it and what type of concurrency you use (threads vs. greenlets). If you want to be on the safe side, then yes, adding a critical section might be a good idea.

**Reply by seatedscribe** [reply]
Thanks a lot for your consistent support
